### DISCUSSION TITLE
I need to modify the value of string type if the property name is "xyz" in JsonConverter<string> using System.Text.Json

### DISCUSSION THREAD

**Discussion by dnxit** [dotnet/runtime#100071] [Q&A]
I don't want to decorate my properties with this converter like

    [JsonConverter(typeof(CustomJsonConverter))]
    public string xyz {get;set;}

I need to use it like this

    JsonSerializerOptions _options = new()
	{
		DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
		PropertyNameCaseInsensitive = true,
		Converters = { new CustomJsonConverter() }
	};

    var result = JsonSerializer.DeserializeAsync<T>(stream, _options);

    public class CustomJsonConverter : JsonConverter<string>
    {
	    public override string Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
	    {
		     if (reader.TokenType == JsonTokenType.PropertyName)
		     {
			     var propertyName = reader.GetString();

 			     if (propertyName == "xyz")
			     {
				      reader.Read(); // Move to the property value
 				      string originalValue = reader.GetString();

				      // Modify the string value (e.g., add a prefix)
				     string modifiedValue = "Modified: " + originalValue;

				     return modifiedValue;
			    }
		}

		return reader.GetString();
	}

	public override void Write(Utf8JsonWriter writer, string value, JsonSerializerOptions options)
	{
		// Write the value as a string
		writer.WriteStringValue(value);
	}
}

reader.TokenType is always equals to JsonTokenType.String so it never goes in that if statement

    if (reader.TokenType == JsonTokenType.PropertyName)

any help would be much appreciated

**Comment by dotnet-policy-service** [comment]
Tagging subscribers to this area: @dotnet/area-system-text-json, @gregsdennis
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.

**Comment by gregsdennis** [comment]
Your JSON converter is only called after the serializer reads the property name and moves on to read the string value.

Typically converters only read the value, not also the property name.

If you want to capture the property name, you'll need a converter for your object type, not `string`.

**Comment by Ilchert** [comment]
Hello @dnxit you can achieve required behavior without converters.
```C#
var o = new JsonSerializerOptions()
{
    TypeInfoResolver = new DefaultJsonTypeInfoResolver().WithAddedModifier(p =>
    {
        foreach (var prop in p.Properties)
            if (prop is { Name: "Prop", Set: { } setter })
            {
                prop.Set = (o, value) =>
                {
                    value = "Modified: " + value;
                    setter(o, value);
                };
            }
    })
};
var a = JsonSerializer.Deserialize<A>("""{"Prop":"str"}""", o);

Console.WriteLine(a.Prop); /// print Modified: str

class A
{
    public string Prop { get; set; }
}
```

**Reply by dnxit** [reply]
I tried it, below is my code but after deserialization items are null 

```
using System.Text.Json;
using System.Text.Json.Serialization.Metadata;


var o = new JsonSerializerOptions()
{
	TypeInfoResolver = new DefaultJsonTypeInfoResolver().WithAddedModifier(p =>
	{
		foreach (var prop in p.Properties)
		{
			if (prop is { Name: "TenantName", Set: { } setter })
			{
				prop.Set = (o, value) =>
				{
					value = "Modified: " + value;
					setter(o, value);
				};
			}
		}
	})
};


var json = "{\r\n  \"items\": [\r\n    {\r\n      \"id\": \"5f71659e-605d-4297-8dc8-d8de7cf2d80f\",\r\n      \"name\": \"Admin\",\r\n      \"tenantName\": \"Root1\"\r\n    },\r\n    {\r\n      \"id\": \"8cd79035-201e-4efb-80e4-b9f6ba9311e4\",\r\n      \"name\": \"Candidate\",\r\n      \"tenantName\": \"Root2\"\r\n    }\r\n  ],\r\n  \"filteredCount\": 2,\r\n  \"totalCount\": 2\r\n}";

var a = JsonSerializer.Deserialize<PagedResult<MyResponse>>(json, o);

Console.WriteLine(a); /// print Modified: str


public class PagedResult<T>
{
	public IReadOnlyList<T> Items { get; set; }

	public int TotalCount { get; set; }
}

public class MyResponse
{
	public string Id { get; set; }

	public string Name { get; set; }

	public string TenantName { get; set; }
}
```
